cltk/phonology/orthophonology.py

```diff
@@ -0,0 +1,477 @@
+'''
+A module for representing the orthophonology of a language: 
+the mapping from orthographic representations to IPA symbols.
+
+Based on many ideas in cltk.phonology.utils by Clément Besnier <clemsciences@gmail.com>.
+'''
+
+from enum import IntEnum, auto
+from copy import deepcopy
+from functools import reduce
+import re
+
+__author__ = ["John Stewart <johnstewart@aya.yale.edu>"]
+
+# ------------------- Phonological Features -------------------
+
+# The list of features and their values are from the IPA charts.
+# Features for non-pulmonic consonants (e.g. clicks, implosives) are not yet provided.
+
+class PhonologicalFeature(IntEnum):
+	pass
+
+class Consonantal(PhonologicalFeature):
+	neg = auto()
+	pos = auto()
+
+class Voiced(PhonologicalFeature):
+	neg = auto()
+	pos = auto()
+
+class Aspirated(PhonologicalFeature):
+	neg = auto()
+	pos = auto()
+
+class Geminate(PhonologicalFeature):
+	neg = auto()
+	pos = auto()
+
+class Roundedness(PhonologicalFeature):
+	neg = auto()
+	pos = auto()
+
+class Length(PhonologicalFeature):
+    short = auto()
+    long = auto()
+    overlong = auto()
+
+# order for Height, Backness, and Manner is important
+# the feature values must be ordered by *increasing sonority*
+class Height(PhonologicalFeature):
+    close = auto()
+    near_close = auto()
+    close_mid = auto()
+    mid = auto()
+    open_mid = auto()
+    near_open = auto()
+    open = auto()
+
+class Backness(PhonologicalFeature):
+    front = auto()
+    central = auto()
+    back = auto()
+
+class Manner(PhonologicalFeature):
+	stop = auto()
+	fricative = auto()
+	affricate = auto()
+	nasal = auto()
+	lateral = auto()
+	trill = auto()
+	spirant = auto()
+	approximant = auto()
+
+class Place(PhonologicalFeature):
+    bilabial = auto()
+    labio_dental = auto()
+    dental = auto()
+    alveolar = auto()
+    post_alveolar = auto()
+    retroflex = auto()
+    palatal = auto()
+    velar = auto()
+    uvular = auto()
+    glottal = auto()
+
+
+# ------------------- Phonemes -------------------
+
+class AbstractPhoneme:

```

Comment:
More generic than AbstractVowel and AbstractConsonant. I like it. Because features are all objects, it's much easier to manage them. However, binary features like Voiced are strange because a boolean would be more natural. But I support your decision.

Reply:
I went back/forth on this.  In the end I felt that having a uniform type made the syntax of logical expressions involving features more uniform.  This is a place where I really miss true types.